File: src/TicTacToe.erl
-module(tic_tac_toe).
-export([start/0, play_game/2]).

start() ->
    io:format("Welcome to Tic-Tac-Toe!~n"),
    io:format("Players use 'x' and 'o'. Enter the number of the cell (1-9) to place your mark.~n"),
    Board = initialize_board(),
    play_game(Board, x).

initialize_board() ->
    lists:seq(1, 9).

display_board(Board) ->
    io:format("~n--- Tic-Tac-Toe Board ---~n"),
    io:format("-------------~n"),
    [B1,B2,B3,B4,B5,B6,B7,B8,B9] = Board,
    io:format("| ~w | ~w | ~w |~n", [B1, B2, B3]),
    io:format("-------------~n"),
    io:format("| ~w | ~w | ~w |~n", [B4, B5, B6]),
    io:format("-------------~n"),
    io:format("| ~w | ~w | ~w |~n", [B7, B8, B9]),
    io:format("-------------~n"),
    io:format("-------------------------~n~n").

make_move(OldBoard, Player, Choice, NewBoard) when is_integer(Choice), Choice >= 1, Choice =< 9 ->
    Pos = Choice - 1,
    case lists:nth(Choice, OldBoard) of
        Cell when is_integer(Cell) ->
            NewBoard = lists:replace_nth(Choice, Player, OldBoard),
            {ok, NewBoard};
        _ ->
            {error, taken}
    end;
make_move(_OldBoard, _Player, _Choice, _NewBoard) ->
    {error, invalid_input}.

check_win(Board, Player) ->
    [B1,B2,B3,B4,B5,B6,B7,B8,B9] = Board,
    (B1 == Player andalso B2 == Player andalso B3 == Player) orelse
    (B4 == Player andalso B5 == Player andalso B6 == Player) orelse
    (B7 == Player andalso B8 == Player andalso B9 == Player) orelse
    (B1 == Player andalso B4 == Player andalso B7 == Player) orelse
    (B2 == Player andalso B5 == Player andalso B8 == Player) orelse
    (B3 == Player andalso B6 == Player andalso B9 == Player) orelse
    (B1 == Player andalso B5 == Player andalso B9 == Player) orelse
    (B3 == Player andalso B5 == Player andalso B7 == Player).

check_draw(Board) ->
    not lists:member_fun(fun is_integer/1, Board).

get_player_choice() ->
    io:format("Enter your move (1-9): "),
    Line = io:get_line(""),
    case string:to_integer(string:trim(Line)) of
        {ok, Choice, _} -> Choice;
        _ -> 0
    end.

play_game(Board, CurrentPlayer) ->
    display_board(Board),
    case check_win(Board, CurrentPlayer) of
        true ->
            io:format("Player ~w wins! Congratulations!~n", [CurrentPlayer]),
            io:format("Game over. Thanks for playing!~n");
        false ->
            case check_draw(Board) of
                true ->
                    io:format("It's a draw!~n"),
                    io:format("Game over. Thanks for playing!~n");
                false ->
                    io:format("Player ~w's turn.~n", [CurrentPlayer]),
                    Choice = get_player_choice(),
                    case make_move(Board, CurrentPlayer, Choice) of
                        {ok, NewBoard} ->
                            NextPlayer = case CurrentPlayer of
                                             x -> o;
                                             o -> x
                                         end,
                            play_game(NewBoard, NextPlayer);
                        {error, taken} ->
                            io:format("Invalid move. That cell is already taken. Please choose an empty cell.~n"),
                            play_game(Board, CurrentPlayer);
                        {error, invalid_input} ->
                            io:format("Invalid input. Please enter a number from 1-9 for an empty cell.~n"),
                            play_game(Board, CurrentPlayer)
                    end
            end
    end.

File: src/TicTacToe.ex
defmodule TicTacToe do
  @moduledoc """
  A simple command-line Tic-Tac-Toe game implemented in Elixir.
  Players take turns entering a number from 1 to 9 to place their mark.
  """

  def initialize_board() do
    1..9 |> Enum.to_list()
  end

  def display_board(board) do
    IO.puts "\n--- Tic-Tac-Toe Board ---"
    IO.puts "-------------"
    [r1, r2, r3] = Enum.chunk_every(board, 3)

    IO.puts ~s(| #{format_cell(r1 |> Enum.at(0))} | #{format_cell(r1 |> Enum.at(1))} | #{format_cell(r1 |> Enum.at(2))} |)
    IO.puts "-------------"
    IO.puts ~s(| #{format_cell(r2 |> Enum.at(0))} | #{format_cell(r2 |> Enum.at(1))} | #{format_cell(r2 |> Enum.at(2))} |)
    IO.puts "-------------"
    IO.puts ~s(| #{format_cell(r3 |> Enum.at(0))} | #{format_cell(r3 |> Enum.at(1))} | #{format_cell(r3 |> Enum.at(2))} |)
    IO.puts "-------------"
    IO.puts "-------------------------\n"
  end

  defp format_cell(val) when is_atom(val), do: String.upcase(Atom.to_string(val))
  defp format_cell(val), do: to_string(val)

  def check_win(board, player) do
    win_conditions = [
      [0, 1, 2], [3, 4, 5], [6, 7, 8], # Rows
      [0, 3, 6], [1, 4, 7], [2, 5, 8], # Columns
      [0, 4, 8], [2, 4, 6]            # Diagonals
    ]

    Enum.any?(win_conditions, fn [i1, i2, i3] ->
      (Enum.at(board, i1) == player) and
      (Enum.at(board, i2) == player) and
      (Enum.at(board, i3) == player)
    end)
  end

  def check_draw(board) do
    Enum.all?(board, fn cell -> not is_integer(cell) end)
  end

  def get_player_choice(current_player) do
    IO.puts ~s(Player #{String.upcase(Atom.to_string(current_player))}'s turn.)
    IO.write "Enter your move (1-9): "
    input = IO.read(:line) |> String.trim()

    case Integer.parse(input) do
      {choice, ""} when choice >= 1 and choice <= 9 -> choice
      _ -> 0
    end
  end

  def make_move(old_board, player, choice) when is_integer(choice) and choice >= 1 and choice <= 9 do
    index = choice - 1
    case Enum.at(old_board, index) do
      cell when is_integer(cell) ->
        new_board = Enum.replace_at(old_board, index, player)
        {:ok, new_board}
      _ ->
        {:error, :taken}
    end
  end
  def make_move(_old_board, _player, _choice), do: {:error, :invalid_input}

  def play_game(board, current_player) do
    display_board(board)

    cond do
      check_win(board, current_player) ->
        IO.puts ~s(Player #{String.upcase(Atom.to_string(current_player))} wins! Congratulations!)
        IO.puts "Game over. Thanks for playing!"
      check_draw(board) ->
        IO.puts "It's a draw!"
        IO.puts "Game over. Thanks for playing!"
      true ->
        choice = get_player_choice(current_player)
        case make_move(board, current_player, choice) do
          {:ok, new_board} ->
            next_player = if current_player == :x, do: :o, else: :x
            play_game(new_board, next_player)
          {:error, :taken} ->
            IO.puts "Invalid move. That cell is already taken. Please choose an empty cell."
            play_game(board, current_player)
          {:error, :invalid_input} ->
            IO.puts "Invalid input. Please enter a number from 1-9 for an empty cell."
            play_game(board, current_player)
        end
    end
  end

  def start() do
    IO.puts "Welcome to Tic-Tac-Toe!"
    IO.puts "Players use 'X' and 'O'. Enter the number of the cell (1-9) to place your mark."
    initial_board = initialize_board()
    play_game(initial_board, :x)
  end
end
